UnitTests/BraintreeVisaCheckoutTests/BTVisaCheckoutClient_Tests.swift

```diff
@@ -1,47 +1,470 @@
-        XCTAssertNil(address.postalCode)
-        XCTAssertNil(address.countryCode)
-        XCTAssertNil(address.phoneNumber)
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let encryptedKey = "b"
+        let encryptedPaymentData = "c"
+
+        let expecation = expectation(description: "Tokenization error due to malformed CheckoutResult")
+        
+        client.tokenize(statusCode: .statusSuccess, callID: nil, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { nonce, error in
+            if nonce != nil {
+                XCTFail("Expected nonce")
+                return
+            }
+            
+            guard let error = error as? NSError else {
+                XCTFail("Error expected to be returned in this tests")
+                return
+            }
+
+            XCTAssertEqual(error as! BTVisaCheckoutError, BTVisaCheckoutError.integration)
+            expecation.fulfill()
+        }
+        waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenStatusCodeIndicatesCancellation_callsCompletionWithNilNonceAndError() {
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expectation = self.expectation(description: "Callback invoked")
+
+        
+        client.tokenize(statusCode: .statusUserCancelled, callID: nil, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { result, error in
+
+            XCTAssertNil(result)
+            XCTAssertEqual(error as! BTVisaCheckoutError, BTVisaCheckoutError.canceled)
+            expectation.fulfill()
+        }
+
+        self.waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenStatusCodeIndicatesError_callsCompletionWithError() {
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expectation = self.expectation(description: "Callback invoked")
+
+        client.tokenize(statusCode: .statusInternalError, callID: callID, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { _, error in
+            guard let error = error as NSError? else {
+                XCTFail("Expected an error to be returned")
+                return
+            }
+            XCTAssertEqual(error.code, BTVisaCheckoutError.checkoutUnsuccessful.rawValue)
+            expectation.fulfill()
+        }
+
+        self.waitForExpectations(timeout: 1, handler: nil)
+    }
+
+    func testTokenize_whenStatusCodeIndicatesCancellation_callsAnalyticsWithCancelled() {
+        let client = BTVisaCheckoutClient(apiClient: mockAPIClient)
+        let expectation = self.expectation(description: "Analytic sent")
+        
+        client.tokenize(statusCode: .statusUserCancelled, callID: callID, encryptedKey: encryptedKey, encryptedPaymentData: encryptedPaymentData) { _, _ in
+            XCTAssertEqual(self.mockAPIClient.postedAnalyticsEvents.last, BTVisaCheckoutAnalytics.tokenizeFailed.description)
```

Comment:
nit to just be a bit more explicit on this assertion:
```suggestion
            XCTAssertTrue(mockAPIClient.postedAnalyticsEvents.contains("EVENT_NAME_HERE"))
```

Reply:
Same in below tests